runtime/src/main/java/org/corfudb/runtime/CheckpointWriter.java

```diff
@@ -148,19 +146,37 @@ public Token appendCheckpoint() {
      */
     @VisibleForTesting
     public Token appendCheckpoint(Token snapshotTimestamp) {
-        appendCheckpointOnSnapshot(snapshotTimestamp);
-        return snapshotTimestamp;
+        return appendCheckpointOnSnapshot(snapshotTimestamp);
     }
 
     /**
      * Write a checkpoint which reflects the state at snapshot.
      *
      * @param snapshotTimestamp snapshot at which the checkpoint is taken.
      */
-    private void appendCheckpointOnSnapshot(Token snapshotTimestamp) {
+    private Token appendCheckpointOnSnapshot(Token snapshotTimestamp) {
         long start = System.currentTimeMillis();
 
-        rt.getAddressSpaceView().read(snapshotTimestamp.getSequence());
+        // If snapshotTimestamp (stream tail) is a negative value
+        // we need to enforce a hole to ensure the start log address
+        // is not negative (as MultiCheckpointWriter will select the min of all checkpoints)
+        // Having a negative value will prevent trimming from happening.
+
+        // Note: we could simply not checkpoint empty streams with the introduction
+        // of address maps, but for now, we'll keep it consistent with old backpointer
+        // strategy, which required checkpoints for empty streams to avoid TrimmedExceptions
+        if (!Address.isAddress(snapshotTimestamp.getSequence())) {
+            snapshotTimestamp = forceHole();
```

Comment:
Is it better to define another type of LogData to represent dummy data but has backpointers? If we still want checkpoint to be compatible with back-pointer mechanism, we should avoid holes if possible.

Reply:
I don't really think we need that, anyways our goal is to completely move away from back pointers. And address maps does not have this issue.